src/backend/distributed/planner/query_pushdown_planning.c

```diff
@@ -1718,45 +1756,61 @@ UpdateColumnToMatchingTargetEntry(Var *column, Node *flattenedExpr, List *target
 	foreach(targetEntryCell, targetEntryList)
 	{
 		TargetEntry *targetEntry = (TargetEntry *) lfirst(targetEntryCell);
+		NodeTag targetEntryType = nodeTag(targetEntry->expr);
+		bool doneProcessing = false;
 
-		if (IsA(targetEntry->expr, Var))
+		switch (targetEntryType)
 		{
-			Var *targetEntryVar = (Var *) targetEntry->expr;
-
-			if (IsA(flattenedExpr, Var) && equal(flattenedExpr, targetEntryVar))
+			case T_Var:
 			{
-				column->varno = 1;
-				column->varattno = targetEntry->resno;
+				Var *targetEntryVar = (Var *) targetEntry->expr;
+
+				if (IsA(flattenedExpr, Var) && equal(flattenedExpr, targetEntryVar))
+				{
+					column->varno = 1;
+					column->varattno = targetEntry->resno;
+					doneProcessing = true;
+				}
 				break;
 			}
-		}
-		else if (IsA(targetEntry->expr, CoalesceExpr))
-		{
-			/*
-			 * flatten_join_alias_vars() flattens full oter joins' columns that is
-			 * in the USING part into COALESCE(left_col, right_col)
-			 */
-			CoalesceExpr *targetCoalesceExpr = (CoalesceExpr *) targetEntry->expr;
 
-			if (IsA(flattenedExpr, CoalesceExpr) && equal(flattenedExpr,
-														  targetCoalesceExpr))
+			case T_CoalesceExpr:
+			case T_FuncExpr:
+
+				/*
+				 * FlattenJoinVars() flattens full oter joins' columns that is
+				 * in the USING part into COALESCE(left_col, right_col)
+				 * It also leaves function calls as they were, but modifies Var references in it.
+				 */
 			{
-				Oid expressionType = exprType(flattenedExpr);
-				int32 expressionTypmod = exprTypmod(flattenedExpr);
-				Oid expressionCollation = exprCollation(flattenedExpr);
-
-				column->varno = 1;
-				column->varattno = targetEntry->resno;
-				column->vartype = expressionType;
-				column->vartypmod = expressionTypmod;
-				column->varcollid = expressionCollation;
+				NodeTag flattenedExprType = nodeTag(flattenedExpr);
+
+				if (flattenedExprType == targetEntryType && equal(flattenedExpr,
+																  targetEntry->expr))
+				{
+					Oid expressionType = exprType(flattenedExpr);
+					int32 expressionTypmod = exprTypmod(flattenedExpr);
+					Oid expressionCollation = exprCollation(flattenedExpr);
+
+					column->varno = 1;
+					column->varattno = targetEntry->resno;
+					column->vartype = expressionType;
+					column->vartypmod = expressionTypmod;
+					column->varcollid = expressionCollation;
+
+					doneProcessing = true;
```

Comment:
Minor readability question: Is there a reason not to return here (or break)? We could get rid of this intermediate variable?

Reply:
returning looks better. since it is inside switch, break (already there) only exits from switch block, and still stay at loop block.